man/fi_profile.3.md

```diff
@@ -0,0 +1,281 @@
+to read variables (e.g, FI_VAR_UNEXP_MSG_CNT) when needed.  It also
+can notify user with a fi_cq_tagged_entry data ("param") which contains 
+<src_addr, tag, size> to indicate how the event occured.
+
+# PROFILE DESCRIPTION
+
+Variables and events are described using struct fi_profile_desc: 
+
+{% highlight c %}
+struct fi_profile_desc {
+	uint32_t id;
+	enum fi_datatype type;
+	size_t size;
+	const char *name;
+	const char *desc;
+};
+{% endhighlight %}
+
+## id
+
+Each variable or event has a unique identifier (id). Providers provide this 
+id and users use it to get the variable value or register callbacks for 
+the event.
+
+## type
+
+The data type of the variable. 
+
+## size
+
+The size (in byte) of the data. "type" and "size" can be used to support 
+any provider-specific data.
+
+## name
+
+Reference to the name of a variable or an event. Storage is allocated and 
+managed by the provider. Variables or events with the same name should have 
+an identical id. 
+
+## desc
+Reference to the description of a variable or an event. Storage is allocated
+and managed by the provider.
+
+# fi_profile_open
+
+fi_profile_open is used to initiate a profiling on a target endpoint or domain 
+specified by the fid. Upon retuning successfully, a fid_profile is returned 
+and can be used to perform profiling operations.  If the target endpoint or 
+domain does not support fabric profiling interface, the open call will fail 
+with error code -FI_ENOSYS.
+
+# fi_profile_close
+
+Close the profiling operations and release related resources.
+
+# fi_profile_query_vars / fi_profile_query_events
+
+fi_profile_query_vars and fi_profile_query_events calls can be used to query 
+available variables or events from the target endpoint or domain associated 
+with the profile object. On input, the buffer ("varlist" or "eventlist") 
+should be allocated, and the "count" indicates the maximum number of entries 
+(variables or events) available in the buffer. On output, the "count" 
+indicates the actual number of entries in the buffer. If the input "count" 
+is smaller than the actual number of entries available, only "count" number 
+of entries will be set in the buffer, and "-FI_ETOOSMALL" will be returned. 
+If the buffer is NULL, upon returning, the "count" indicates the total number 
+of entries available in the target endpoint or domain. User can allocate 
+space and call the function again.
+
+# fi_profile_reg_callback
+
+fi_profile_reg_callback registers a callback function to the event. The event 
+should be in the event list obtained using the fi_profile_query_events, 
+otherwise, an error code -FI_EINVAL will be returned. 
+
+## Callback
+
+It is used to receive notifications from a provider when an event occurs. 
+It is defined by the user and called by a provider as an additional operation,
+normally before or after data transfer.  This will increase the overhead of 
+the operation.  Callback functions should avoid having blocking calls, or 
```

Comment:
A nit, but I think it would be preferable to go with the semantic that "Callback functions must avoid blocking calls...". Time spent in this callback can be preventing hardware queues from progressing.